session-replay/Sources/DatadogSessionReplay/DatadogCoreIntegration/RequestBuilder/JSON/SegmentJSONBuilder.swift

```diff
@@ -0,0 +1,108 @@
+/*
+ * Unless explicitly stated otherwise all files in this repository are licensed under the Apache License Version 2.0.
+ * This product includes software developed at Datadog (https://www.datadoghq.com/).
+ * Copyright 2019-2020 Datadog, Inc.
+ */
+
+import Foundation
+
+internal struct SegmentJSONBuilder {
+    let source: SRSegment.Source
+
+    /// Groups records into segments.
+    ///
+    /// It scans succeeding records and groups together ones that share the same RUM context. Each group
+    /// stands for a separate segment. To ensure it produces an optimal number of segments, the `records`
+    /// array must be ordered in a way that the same RUM context appears in continuous ranges.
+    ///
+    /// - Parameter records: records to bundle within segments
+    /// - Returns: array of serializable segments
+    func createSegmentJSONs(from records: [EnrichedRecordJSON]) -> [SegmentJSON] {
+        guard !records.isEmpty else {
+            return []
+        }
+
+        var segments: [SegmentJSON?] = []
+        var current = records[0]
+
+        // Current segment state:
+        var startIndex = 0
+        var hasFullSnapshot = current.hasFullSnapshot
+        var startTime = current.earliestTimestamp
+        var endTime = current.latestTimestamp
+
+        // Loop through the array to find continuous ranges of records that have
+        // the same RUM context. For each range create a segment:
+        for index in (1..<records.count) {
```

Comment:
This could be done fancier with functional operators, however `createSegmentJSONs(from)` is part of the critical path in SR uploads and needs to loop through all batched events, so I wanted to keep it fast and performant.